scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java

```diff
@@ -104,28 +92,19 @@ private <T extends ScimResource> T apply(final T source, final PatchOperation pa
     if (valuePathExpression.getAttributeExpression() != null && attributeObject instanceof Collection<?>) {
       // apply expression filter
       Collection<Object> items = (Collection<Object>) attributeObject;
-      Collection<Object> updatedCollection = items.stream().map(item -> {
-        KeyedResource keyedResource = (KeyedResource) item;
-        Map<String, Object> keyedResourceAsMap = objectAsMap(item);
-        Predicate<KeyedResource> pred = FilterExpressions.inMemory(valuePathExpression.getAttributeExpression(), baseSchema);
-        if (pred.test(keyedResource)) {
+      items.forEach(item -> {
+        Predicate<Object> pred = FilterExpressions.inMemory(valuePathExpression.getAttributeExpression(), baseSchema);
+        if (pred.test(item)) {
           String subAttributeName = valuePathExpression.getAttributePath().getSubAttributeName();
```

Comment:
One thing we may need to consider with either of these solutions is that they assume a single nested value. `email.types`, `address.city`, etc.

IIRC, all of the official spec schemas only have one level like this, but it's possible there could be a more complex value in the future or through a custom extension.

That said, I'm mostly just adding a note to think about it; I don't think it needs to block this getting merged (the `develop` branch just throws an exception, so this is a huge improvement and probably hits 90% of cases)